### PR TITLE
EID-312 Enable European Users

### PR DESCRIPTION
**Pull request by lamerexter**
- This is a re-application of EID-293 to reduce complexity of the MatchingServiceResource via a simple delegation model.
- HealthCheck, Verify AttributeQuery handling and Eidas AttributeQuery handling are all given their own single-purpose service (think 12-factor...)
- Result is that MatchingServiceResource::receiveSoapRequest method is reduced to 3 lines...
- All tests run green, as expected.

### CODE REVIEW

File: src/main/java/uk/gov/ida/matchingserviceadapter/MatchingServiceAdapterModule.java [hunk 1]
```diff
@@ -130,6 +149,88 @@ protected void configure() {
...
+        @Named("CountryCertificateValidator") Optional<CertificateValidator> countryCertificateValidator,
+        X509CertificateFactory x509CertificateFactory,
+        MatchingServiceAdapterConfiguration configuration,
+        AssertionDecrypter assertionDecrypter
+    ) {
+        return countryMetadataResolver.map(cmr ->
+            new EidasMatchingService(
+                new EidasAttributeQueryValidator(
+                    verifyMetadataResolver,
+                    cmr,
```
Comment (lamerexter) [inline src/main/java/uk/gov/ida/matchingserviceadapter/MatchingServiceAdapterModule.java]: This is rather short for countryMetadataResolver

**Comment by thomasalanlee** [comment]
This is too large. Perhaps introducing the ServiceLocator could be a separate change

The commit messages are not descriptive enough. If you are introducing a significant change (e.g. the ServiceLocator) please describe why in the commit messages.

There are quite a few places acronyms or abbreviations have been used. We should favour clarity over brevity (e.g. ctor, cmr)

File: src/test/java/uk/gov/ida/matchingserviceadapter/services/EidasMatchingServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
+package uk.gov.ida.matchingserviceadapter.services;
+
+import org.beanplanet.messages.domain.Message;
```
Comment (lamerexter) [inline src/test/java/uk/gov/ida/matchingserviceadapter/services/EidasMatchingServiceTest.java]: Don't worry: all references to the Beanplanet library HAVE BEEN REMOVED in the next commit changeset!

**Comment by adityapahuja** [comment]
Test this please.

**Comment by lamerexter** [comment]
Thanks @thomasalanlee.  I have addressed all of your review items:

1. - Beanplanet library references - all were already removed by the next commit in the list.
2. - PR Job build failure - Jenkins is being restarted
3. - Commit messages - messages made more descriptive.
4. - Acronyms - ctor and cmr renamed accordingly.

Thanks.

**Comment by adityapahuja** [comment]
Test this please.

**Comment by adityapahuja** [comment]
Test this please.

**Comment by TimWSpence** [comment]
test this please

**Comment by TimWSpence** [comment]
test this please

**Comment by adityapahuja** [comment]
Test this please.

**Comment by lamerexter** [comment]
Test this please.

File: src/test/java/uk/gov/ida/matchingserviceadapter/validators/datetime/JodaAndJavaDateTimeConvertersTest.java [hunk 3]
```diff
@@ -16,7 +16,7 @@
 
 public class JodaAndJavaDateTimeConvertersTest {
     @Test
-    public void ctorForTestCoverageOnly() {
+    public void constructorForTestCoverageOnly() {
```
Comment (thomasalanlee) [inline src/test/java/uk/gov/ida/matchingserviceadapter/validators/datetime/JodaAndJavaDateTimeConvertersTest.java]: I'm not sure we need this test do we?

File: src/test/java/uk/gov/ida/matchingserviceadapter/domain/VerifyMatchingServiceResponseTest.java [hunk 4]
```diff
@@ -9,7 +9,7 @@
 
 public class VerifyMatchingServiceResponseTest {
     @Test
-    public void ctor() {
+    public void constructor() {
```
Comment (thomasalanlee) [inline src/test/java/uk/gov/ida/matchingserviceadapter/domain/VerifyMatchingServiceResponseTest.java]: Please describe what behaviour the test is verifying in the test title.

File: src/test/java/uk/gov/ida/matchingserviceadapter/resources/DelegatingMatchingServiceResponseRendererTest.java [hunk 5]
```diff
@@ -19,7 +19,7 @@
     public final ExpectedException exception = ExpectedException.none();
 
     @Test
-    public void ctor() {
+    public void constructor() {
```
Comment (thomasalanlee) [inline src/test/java/uk/gov/ida/matchingserviceadapter/resources/DelegatingMatchingServiceResponseRendererTest.java]: Please describe the behaviour we're verifying in the test name

File: src/test/java/uk/gov/ida/matchingserviceadapter/services/EidasMatchingServiceTest.java [hunk 6]
```diff
@@ -26,7 +26,7 @@
 
     @SuppressWarnings("unchecked")
     @Test
-    public void ctor() {
+    public void constructor() {
```
Comment (thomasalanlee) [inline src/test/java/uk/gov/ida/matchingserviceadapter/services/EidasMatchingServiceTest.java]: Please describe the behaviour we're verifying in the test name

File: src/test/java/uk/gov/ida/matchingserviceadapter/validators/IssueInstantValidatorTest.java [hunk 7]
```diff
@@ -29,7 +29,7 @@
     private static final Duration CLOCK_DELTA = Duration.parse("PT1M");
 
     @Test
-    public void ctorTtlFutureMessageDateTimeProviderAndDurations() {
+    public void constructorWithTtlFutureMessageDateTimeProviderAndDurations() {
```
Comment (thomasalanlee) [inline src/test/java/uk/gov/ida/matchingserviceadapter/validators/IssueInstantValidatorTest.java]: Please describe the behaviour we're verifying in the test name

File: src/test/java/uk/gov/ida/matchingserviceadapter/validators/datetime/DurationNotExceededValidatorTest.java [hunk 8]
```diff
@@ -23,7 +23,7 @@
     public static final MessageImpl DEFAULT_MESSAGE = globalMessage("theCode", "theMessage");
 
     @Test
-    public void ctorMessageFromToAndMaximum() {
+    public void constructorMessageFromToAndMaximum() {
```
Comment (thomasalanlee) [inline src/test/java/uk/gov/ida/matchingserviceadapter/validators/datetime/DurationNotExceededValidatorTest.java]: Please describe the behaviour we're verifying in the test name

File: src/test/java/uk/gov/ida/matchingserviceadapter/validators/datetime/DurationNotExceededValidatorTest.java [hunk 9]
```diff
@@ -41,7 +41,7 @@ public void ctorMessageFromToAndMaximum() {
     }
 
     @Test
-    public void ctorMessageFromAndMaximum() {
+    public void constructorMessageFromAndMaximum() {
```
Comment (thomasalanlee) [inline src/test/java/uk/gov/ida/matchingserviceadapter/validators/datetime/DurationNotExceededValidatorTest.java]: Please describe the behaviour we're verifying in the test name

File: src/test/java/uk/gov/ida/matchingserviceadapter/validators/datetime/NotInTheFutureValidatorTest.java [hunk 10]
```diff
@@ -21,7 +21,7 @@
     public static final MessageImpl DEFAULT_MESSAGE = globalMessage("theCode", "theMessage");
 
     @Test
-    public void ctorMessageAndAtProvider() {
+    public void constructorMessageAndAtProvider() {
```
Comment (thomasalanlee) [inline src/test/java/uk/gov/ida/matchingserviceadapter/validators/datetime/NotInTheFutureValidatorTest.java]: Please describe the behaviour we're verifying in the test name

File: src/main/java/uk/gov/ida/matchingserviceadapter/services/DelegatingMatchingService.java [hunk 11]
```diff
@@ -31,13 +31,18 @@ public DelegatingMatchingService(ServiceLocator<MatchingServiceRequestContext, M
...
-
     @Override
     public MatchingServiceResponse handle(MatchingServiceRequestContext requestContext) {
         AttributeQuery attributeQuery = unwrapAttributeQuery(requestContext.getAttributeQueryDocument());
         requestContext.setAttributeQuery(attributeQuery);
         requestContext.setAssertions(decryptAssertions(attributeQuery));
-        return matchingServiceLocator.findServiceFor(requestContext).handle(requestContext);
+
+        MatchingService delegateService = matchingServiceLocator.findServiceFor(requestContext);
+        if (delegateService == null) {
```
Comment (thomasalanlee) [inline src/main/java/uk/gov/ida/matchingserviceadapter/services/DelegatingMatchingService.java]: We should probably use Optional.orElseThrow() to handle null checks

File: src/test/java/uk/gov/ida/matchingserviceadapter/domain/VerifyMatchingServiceResponseTest.java [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+import org.junit.Test;
+import uk.gov.ida.matchingserviceadapter.saml.transformers.outbound.OutboundResponseFromMatchingService;
+
+import static org.hamcrest.CoreMatchers.sameInstance;
+import static org.junit.Assert.assertThat;
+import static org.mockito.Mockito.mock;
+
+public class VerifyMatchingServiceResponseTest {
+    @Test
+    public void ctor() {
```
Comment (thomasalanlee) [inline src/test/java/uk/gov/ida/matchingserviceadapter/domain/VerifyMatchingServiceResponseTest.java]: Please rename this test to describe what behaviour it is verifying (or delete if we don't need it)

File: src/test/java/uk/gov/ida/matchingserviceadapter/resources/MatchingServiceResourceTest.java [hunk 13]
```diff
@@ -24,7 +28,9 @@
 import java.util.jar.Attributes;
 
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.hamcrest.CoreMatchers.sameInstance;
```
Comment (thomasalanlee) [inline src/test/java/uk/gov/ida/matchingserviceadapter/resources/MatchingServiceResourceTest.java]: Please switch to use AssertJ

**Pull request merged**